### ISSUE TITLE
Sudden spike in proxy memory usage 

### ISSUE THREAD

**Issue by klehigh** [zeek/zeek#894]
On 3.1.1 we are seeing a sudden spike in proxy memory usage, from ~ 2GB to 35GB in about 20 minutes.  This appears to happen when we have a short high-volume burst of network activity.  In addition, if known_services use_service_store is true, we see known_services start to log services on every connection (resulting in millions of entries for a DNS server, for instance).  If using use_service_store=F, we still see the explosive memory growth, but known_services continues to log normally.

I have two proxy instances configured, but we've observed this problem with higher numbers of proxies as well (up to five).

I'm running on FreeBSD.  We aren't able to run 3.0.x on this cluster, so I don't know if this problem exists in that branch.

I don't see any relevant errors in reporter.log or std(err|out).log for the proxy instances.

**Comment by jsiwek** [comment]
No immediate clue, but some explanations/ideas that may lead to further insight:

> if known_services use_service_store is true, we see known_services start to log services on every connection (resulting in millions of entries for a DNS server, for instance).

The `service_store_timeout` option defaults to 15 seconds and any timeout leads to just going ahead to log the service since the purpose of that lookup-that-timed-out was to find out if it was unique or not.  The timeout could be increased, but generally trying to centralize all this data may not be too scalable in your case.

> If using use_service_store=F, we still see the explosive memory growth, but known_services continues to log normally.

With `use_service_store=T`, the store would be centrally located on the manager.  I expect proxies to only be involved with `use_service_store=F`.  So if you see increased proxy memory usage either way, it's potentially something other than just the known-services script contributing.  Other usages of proxies look like only Software, FTP, and IRC (and the other "known" scripts for hosts/certs if you changed their default option from using data stores).

> I have two proxy instances configured, but we've observed this problem with higher numbers of proxies as well (up to five).

Scaling up number of proxies should actually decrease how much gets stored in any individual proxy, so I'd hope that could be a way to actually produce smaller peaks per node that aren't so problematic.

Maybe another thing to do is try to identify where the usage is actually coming from.  A crude way to do that could be disable all the major contributors to proxy usage and then bring back in one at a time:

```zeek
redef Software::asset_tracking = NO_HOSTS;
redef Known::service_tracking = NO_HOSTS;
redef Known::host_tracking = NO_HOSTS;
redef Known::cert_tracking = NO_HOSTS;
```

(I don't know for those last two whether you had changed the associated `use_XXX_store` option to toggle between data store versus proxy usage).

**Comment by klehigh** [comment]
With two proxies enabled, we had another instance of proxy instability.  We don't run known-certs, but I disabled all the other items you noted above and we continued to see sudden growth in proxy memory usage, with both proxies growing to ~ 35GB in size, in about 20 minutes.  While troubleshooting, I noticed the simple-scan  Zeek package generates a fairly high volume of proxy related messages, so I disabled that package.  After restart, with that package disabled and no known/software tracking enabled, both proxies grew to ~ 5GB in size (which is about double normal observed size), but stabilized at that point.  When we're able to do more testing, I'll try re-enabling the known/software scripts and experiment with service_store settings and comment in here.

**Comment by JustinAzoff** [comment]
Yeah, simple-scan will generate a message to a proxy for each scan attempt until the scan threshold is crossed.. but it does load balance perfectly across the proxies, so running more would help.

If your individual proxy process cpu usage is 100% you'll want to run more of them so they have a chance to keep up with the traffic.  From what I recall about your network and cluster size I'd expect you to be running 5+ proxies, not just 2.

If you disable encryption in broker(`Broker::disable_ssl=T;`) you can use a simple tcpdump -A command to see what the events being sent to the proxies are.  It's usually pretty easy to see what is causing a proxy to be overloaded just by looking at the strings in the packets.

**Comment by klehigh** [comment]
We've had stability problems from 2 to 5 proxies.  Because the conditions to trigger this are hard to recreate & happen infrequently, it is taking a fair amount of time to eliminate possibilities.

We're not seeing cpu usage issues.  This is entirely memory related, and the spike in memory growth happens in 10 minutes or so (from 500MB in size to 60GB).  Removing simple-scan has helped some, but it hasn't completely resolved the issue.

Disabling broker encryption has been helpful in pointing out issues (like the simple-scan impact), so we'll continue to use that as we dig further into this.

**Comment by JustinAzoff** [comment]
weird.. I'm not sure how  simple-scan would ever use that much ram, unless it's just the internal broker/caf queues buffering that much data:

* simple-scan only tracks info per scanner source ip
* the amount of info tracked per scanner is capped to whatever `scan_threshold` is
* once `scan_threshold` is crossed that info is flushed and only the ip is tracked under `known_scanners`.

This is also perfectly distributed across the proxies, so even if it was using 30GB on a proxy somehow, if you doubled the number from 2 proxies to 4, each proxy would then drop to 15GB each.

You can also try enabling jemalloc profiling for the proxy processes, I have a zeekctl plugin for easily enabling it here https://github.com/JustinAzoff/zeek-jemalloc-profiling. you might have to rebuild jemalloc with profiling enabling first.

**Comment by klehigh** [comment]
The cause of this instability turned out to be overly large FTP data being shared via broker.  This commit https://github.com/zeek/zeek/commit/7e9a3e1e00cd327753914d0071086b60f052fc9d resolved the issue.  Closing this out.  Thanks for all the help in resolving this.
